### DISCUSSION TITLE
Understanding the IExceptionHandler in .NET 8 for Global Exception Handling

### DISCUSSION THREAD

**Discussion by hptruc** [dotnet/aspnetcore#54613] [General] [answered]
Hello everyone,

I've been using a custom middleware approach for handling global exceptions in my .NET applications, which involves creating a new middleware and injecting it into the middleware pipeline using the UseMiddleware method as shown below:

```c#
// ExceptionHandlingMiddleware.cs
public class ExceptionHandlingMiddleware(RequestDelegate next)
{
    public async Task Invoke(HttpContext context)
    {
        try
        {
            await next(context);
        }
        catch (Exception ex)
        {
            await HandleExceptionAsync(context, ex);
        }
    }

    private static Task HandleExceptionAsync(HttpContext context, Exception exception)
    {   
        // Process exception here...
    }
}

// Program.cs
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();
app.UseMiddleware<ExceptionHandlingMiddleware>();
```
However, I noticed that .NET 8 introduces a new feature for handling global exceptions, IExceptionHandler. This interface seems to offer a structured way to handle exceptions globally, but I'm not entirely clear on the advantages or the key differences it brings compared to the traditional middleware approach. The basic usage of IExceptionHandler is as follows:

```c#
public class GlobalExceptionHandler : IExceptionHandler
{
    public async ValueTask<bool> TryHandleAsync(HttpContext httpContext, Exception exception, CancellationToken cancellationToken)
    {
        // Process exception here...
    }
}

// Program.cs
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddExceptionHandler<GlobalExceptionHandler>();
var app = builder.Build();
app.UseExceptionHandler(option => {});
```

Could someone explain the key benefits or differences of using IExceptionHandler over the traditional middleware approach for global exception handling? Specifically, I'm interested in understanding any improvements in performance, ease of use, flexibility, or any other aspects that IExceptionHandler might offer.

Thank you in advance for your insights!

**Comment by kimbell** [comment, marked as answer]
I've been where you are; we created custom middleware and a bunch of extensibility options. We are now in the process of throwing that out and replacing it with the standard middleware and IExceptionHandler. 

There are multiple reasons for why we are doing this.
- Less code we need to maintain
- Can rely on standard MS documentation
- Middleware adds metrics
https://github.com/dotnet/aspnetcore/blob/94656557027404804fa57bdbab4bdb90a6169d43/src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddlewareImpl.cs#L126
- Middleware has handling of OperationCanceledException and IOException
https://github.com/dotnet/aspnetcore/blob/94656557027404804fa57bdbab4bdb90a6169d43/src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddlewareImpl.cs#L117
- If your exception handler returns 404 for an exception, you have to manually enable this possibility. 
https://github.com/dotnet/aspnetcore/blob/94656557027404804fa57bdbab4bdb90a6169d43/src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerOptions.cs#L40 

When MS makes improvements, we get it for free.
